src/alignment/pairwise/myers_miller.rs

```diff
@@ -0,0 +1,1042 @@
+        let xlen = x.len();
+        let operations = self.compute_recursive(
+            x,
+            &y[ystart..yend],
+            xlen,
+            ylen,
+            self.scoring.gap_open,
+            self.scoring.gap_open,
+        );
+        // (doesn't need to insert Yclip)
+
+        // Set the clip penalties to the original values
+        self.scoring.xclip_prefix = clip_penalties[0];
+        self.scoring.xclip_suffix = clip_penalties[1];
+        self.scoring.yclip_prefix = clip_penalties[2];
+        self.scoring.yclip_suffix = clip_penalties[3];
+
+        Alignment {
+            score,
+            xstart: 0,
+            xend: xlen,
+            ystart,
+            yend,
+            xlen: x.len(),
+            ylen: y.len(),
+            operations,
+            mode: AlignmentMode::Semiglobal,
+        }
+    }
+    /// Recursively compute alignments of sub-sequences and concatenating them.
+    ///
+    /// - Space complexity: $O(n)$. Precisely, about $128n + log\_2{m}$, where $m =$ `x.len()` and $n =$ `y.len()`.
+    ///   $128n$ is used for four `Vec<i32>` in `find_mid()` and $log\_2{m}$ is due to the activation stack (the
+    ///   latter can be negligible).
+    ///
+    /// # Panics
+    ///
+    /// Rust has a [`recursion_limit` which defaults to 128](https://doc.rust-lang.org/reference/attributes/limits.html#the-recursion_limit-attribute),
+    /// which means the length of the sequence `x` should not exceed $2^128 = 3.4\times10^38$ (well, you'll
+    /// never encounter such gigantic biological sequences)
+    fn compute_recursive(
+        &self,
+        x: TextSlice<'_>,
+        y: TextSlice<'_>,
+        m: usize,
+        n: usize,
+        tb: i32,
+        te: i32,
+    ) -> Vec<AlignmentOperation> {
+        // * m = x.len(); n = y.len()
+        if n == 0 {
+            return vec![AlignmentOperation::Ins; m];
+        }
+        if m == 0 {
+            return vec![AlignmentOperation::Del; n];
+        }
+        if m == 1 {
+            return self.nw_onerow(x[0], y, n, tb, te);
+        }
+        let (imid, jmid, join_by_deletion) = self.find_mid(x, y, m, n, tb, te); // `find_mid()` uses 128n bits
+        return if join_by_deletion {
+            // y.len() === (&y[..jmid]).len() + (&y[jmid..]).len()
+            // so the sum of the space used by the two subtasks is still 128n (n = y.len())
+            let (a, b) = rayon::join(
+                || self.compute_recursive(&x[..imid - 1], &y[..jmid], imid - 1, jmid, tb, 0),
+                || {
+                    self.compute_recursive(
+                        &x[imid + 1..],
+                        &y[jmid..],
+                        m - imid - 1,
+                        n - jmid,
+                        0,
+                        te,
+                    )
+                },
+            );
+            [a, vec![AlignmentOperation::Ins; 2], b].concat()
+        } else {
+            let (a, b) = rayon::join(
```

Comment:
as a user of this library, I generally want to be able to control when algorithms parallelize themselves.  For example if I'm processing a large number of sequences, and this algorithm is one step of many that I need to apply to each read, then I want to parallelize over reads, and switch off parallelization of this algorithm.  Perhaps the `Aligner` struct should have `parallelize` field that controls this?  If we also add a `rayon` feature, then there's a design question of how the parallelization flag interacts with the feature. With a bit of work we could design a scheme that can statically prevent use of the parallel algo unless `rayon` is enabled, a simpler scheme could just panic at runtime ifthe parallel flag is set but `rayon` isn't enabled.

Reply:
I agree. I'll be working on this. In addition, I found a technique used by [Ngila](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4739816) that improves speed of this algorithm without parallelism, which may be useful to implement:

> Ngila starts with the divide-and-conquer method but switches to the holistic method for subsequences smaller than a user-established threshold. This improves its speed without substantially increasing memory requirements.